### PR TITLE
Modernpython

### PR DESCRIPTION
**Pull request by guillaume-alliander**
Adding Python 3.9+ langpack, with pydantic dataclasses.

Those classes cannot be used directly with cimpy, but they do simplify quite a lot of coding. The class tree crawling is built into Base, profiles to which an attribute belongs are added to the attribute definition itself, json dump/parsing is built in. 

Sonarcloud complains about code smells - those are logical, but fixing those would cause discrepancy with the style of the rest of the code, so I leave them as is for now. If that's not the best option, let me know.

### CODE REVIEW

**Comment by chicco785** [comment]
@guillaume-alliander interesting work!

I am working on something similar (albeit I am not using dataclass but BaseModel), and found out a few issues with recursive models across multiple files, which lead me to generate a single file:

https://github.com/zaphiro-technologies/cimgen/pull/5

did you test the generated code?

happy to help :)

File: modernpython/langPack.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+def _get_type_and_default(text, renderer) -> tuple[str, str]:
+    result = renderer(text)
+    # the field {{dataType}} either contains the multiplicity of an attribute if it is a reference or otherwise the
+    # datatype of the attribute. If no datatype is set and there is also no multiplicity entry for an attribute, the
+    # default value is set to None. The multiplicity is set for all attributes, but the datatype is only set for basic
+    # data types. If the data type entry for an attribute is missing, the attribute contains a reference and therefore
+    # the default value is either None or [] depending on the multiplicity. See also write_python_files
+    # The default will be copied as-is, hence the possibility to have default or default_factory.
+    if result in ["M:1", "M:0..1", "M:1..1", ""]:
+        return ("Optional[str]", "default=None")
```
Comment (chicco785) [inline modernpython/langPack.py:67]: If think I understand why you don't have recursion errors: you are not using other dataclassess as attribute type.

Albeit this avoids any recursion and make validation faster, it does not exploit the full power of pydantic, and makes difficult to navigate the data.

**Comment by guillaume-alliander** [comment]
@chicco785 

I based this a lot on the current python serialization. As the recursion aspect was not there, it's not present in my classes :)

You are correct that I do not use other dataclasses as attributes. It would be nice indeed. I am currently heavily using the generated classes, and adding dataclasses as possible attributes could have very interesting aspects. My final goal would be to load the graph in rdflib (to export to xml, json-ld, ttl... for free), and that could help.

**Comment by chicco785** [comment]
> @chicco785
> 
> I based this a lot on the current python serialization. As the recursion aspect was not there, it's not present in my classes :)
>

to keep simple and the parse easy, may make completely sense, it is just more complex to get navigate a linked object (you will have to search it by id through the full list of objects)

>
> You are correct that I do not use other dataclasses as attributes. It would be nice indeed. I am currently heavily using the generated classes, and adding dataclasses as possible attributes could have very interesting aspects. My final goal would be to load the graph in rdflib (to export to xml, json-ld, ttl... for free), and that could help.

of course we want also to be able to import and export to different format, plus ideally make easy to manipulate / filter the objects via APIs (that's why we have been looking into pydantic and we tested the recursion support), beyond that, for us, also storage in a database (so adding also SQL Alchemy support) is an interesting feature.

Maybe we can have a short call, exchange ideas, and see if there is a chance for a collaboration?

**Review by chicco785 — APPROVED** [review]

**Comment by guillaume-alliander** [comment]
@iripiri Would it be possible to get a review for this PR?

More updates here will come, but this PR is a nicely self-contained.

**Comment by iripiri** [comment]
Sorry for the delay, the changes look good to me.
Thanks for your efforts!

**Review by iripiri — APPROVED** [review]
LGTM

**Pull request merged**